File: registry/src/Xs.Registry.Db/Shared/Models/UserMetaPackageAccess.cs
using System;
using System.Collections.Generic;
using System.Linq;

namespace Xs.Registry.Db.Shared
{
    public struct UserMetaPackageAccess
    {
        public bool IsOwner { get; }

        public bool IsWorld { get; }

        private readonly Permission permission;

        internal UserMetaPackageAccess(
            Guid userId,
            Guid ownerId,
            IEnumerable<MetaPackagePermission> permissions
        )
        {
            var category = ownerId == userId ? PermissionCategory.Owner : PermissionCategory.World;

            IsOwner = category == PermissionCategory.Owner;
            IsWorld = category == PermissionCategory.World;

            permission = permissions.FirstOrDefault(p => p.Category == category)?.Permission ?? Permission.None;
        }

        public bool Has(Permission permission) => this.permission.HasFlag(permission);
    }
}

File: registry/site/src/site/src/models/view/UserMetaPackageAccess.ts
import { MetaPackagePermission } from './MetaPackagePermission'
import { Permission } from './Permission'
import { PermissionCategory } from './PermissionCategory'

export class UserMetaPackageAccess {
  public readonly isOwner: boolean
  public readonly isWorld: boolean
  public readonly permission: Permission

  constructor(
    userId: string,
    ownerId: string,
    permissions: MetaPackagePermission[],
  ) {
    const category = ownerId === userId ? PermissionCategory.Owner : PermissionCategory.World

    this.isOwner = category === PermissionCategory.Owner
    this.isWorld = category === PermissionCategory.World

    this.permission = permissions.find(p => p.category === category)!.permission
  }

  public has(permission: Permission) {
    return (this.permission & permission) === permission
  }
}
